vispy/visuals/graphs/layouts/force_directed.py

```diff
@@ -0,0 +1,222 @@
+                    self.optimal * self.optimal/distance**2 -
+                    adjacency_mat*distance/self.optimal
+                )
+            ).sum(axis=1)
+
+            # update positions
+            length = np.sqrt((displacement**2).sum(axis=1))
+            length = np.where(length < 0.01, 0.1, length)
+            delta_pos = displacement * t / length[:, np.newaxis]
+            pos += delta_pos
+            pos = rescale_layout(pos)
+
+            # cool temperature
+            t -= dt
+
+            # Calculate edge vertices and arrows
+            line_vertices, arrows = straight_line_vertices(adjacency_mat,
+                                                           pos, directed)
+
+            yield pos, line_vertices, arrows
+
+    def _sparse_fruchterman_reingold(self, adjacency_mat, directed=False):
+        # Optimal distance between nodes
+        if self.optimal is None:
+            self.optimal = 1 / np.sqrt(self.num_nodes)
+
+        # Change to list of list format
+        # Also construct the matrix in COO format for easy edge construction
+        adjacency_mat = adjacency_mat.tolil()
+        adjacency_coo = adjacency_mat.tocoo()
+
+        if self.pos is None:
+            # Random initial positions
+            pos = np.asarray(
+                np.random.random((self.num_nodes, self.dim)),
+                dtype='f32'
+            )
+        else:
+            pos = self.pos.astype('f32')
+
+        # Yield initial positions
+        line_vertices, arrows = straight_line_vertices(adjacency_coo, pos,
+                                                       directed)
+        yield pos, line_vertices, arrows
+
+        # The initial "temperature"  is about .1 of domain area (=1x1)
+        # This is the largest step allowed in the dynamics.
+        t = 0.1
+        # Simple cooling scheme.
+        # Linearly step down by dt on each iteration so last iteration is
+        # size dt.
+        dt = t / float(self.iterations+1)
+
+        displacement = np.zeros((self.dim, self.num_nodes))
+        for iteration in range(self.iterations):
+            displacement *= 0
+            # Loop over rows
+            for i in range(adjacency_mat.shape[0]):
+                # difference between this row's node position and all others
+                delta = (pos[i]-pos).T
+                # distance between points
+                distance = np.sqrt((delta**2).sum(axis=0))
+                # enforce minimum distance of 0.01
+                distance = np.where(distance < 0.01, 0.01, distance)
+                # the adjacency matrix row
+                row = adjacency_mat.getrowview(i).toarray()
+
+                # displacement "force"
+                displacement[:, i] += (
+                    delta * (
+                        self.optimal *
+                        self.optimal/distance**2 -
+                        row*distance/self.optimal
+                    )
+                ).sum(axis=1)
```

Comment:
I suspect you can boil this down to a `dot` operation between your sparse matrix and an array of `deltas`. This will be fastest if `adjacency_mat` is a `csr` matrix, I think.